crates/fuel-core/src/graphql_api/storage/coins.rs

```diff
@@ -8,21 +8,253 @@ use fuel_core_storage::{
+        write!(
+            f,
+            "retryable_flag={}, owner={}, asset_id={}, amount={}",
+            self.retryable_flag(),
+            self.owner(),
+            self.asset_id(),
+            self.amount()
+        )
+    }
+}
+
+impl TryFrom<&CoinsToSpendIndexKey> for fuel_tx::UtxoId {
+    type Error = ();
+
+    fn try_from(value: &CoinsToSpendIndexKey) -> Result<Self, Self::Error> {
+        let bytes: [u8; COIN_FOREIGN_KEY_LEN] =
+            value.foreign_key_bytes().try_into().map_err(|_| ())?;
+
+        let (tx_id_bytes, output_index_bytes) = bytes.split_at(TxId::LEN);
+        let tx_id = TxId::try_from(tx_id_bytes).map_err(|_| ())?;
+        let output_index =
+            u16::from_be_bytes(output_index_bytes.try_into().map_err(|_| ())?);
+        Ok(fuel_tx::UtxoId::new(tx_id, output_index))
+    }
+}
+
+impl TryFrom<&CoinsToSpendIndexKey> for fuel_types::Nonce {
+    type Error = ();
+
+    fn try_from(value: &CoinsToSpendIndexKey) -> Result<Self, Self::Error> {
+        let bytes: [u8; MESSAGE_FOREIGN_KEY_LEN] =
+            value.foreign_key_bytes().try_into().map_err(|_| ())?;
+        Ok(fuel_types::Nonce::from(bytes))
+    }
+}
+
+impl CoinsToSpendIndexKey {
+    pub fn from_coin(coin: &Coin) -> Self {
+        let retryable_flag_bytes = NON_RETRYABLE_BYTE;
+        let address_bytes = coin.owner.as_ref();
+        let asset_id_bytes = coin.asset_id.as_ref();
+        let amount_bytes = coin.amount.to_be_bytes();
+        let utxo_id_bytes = utxo_id_to_bytes(&coin.utxo_id);
+
+        Self(
+            retryable_flag_bytes
+                .iter()
+                .chain(address_bytes)
+                .chain(asset_id_bytes)
+                .chain(amount_bytes.iter())
+                .chain(utxo_id_bytes.iter())
+                .copied()
+                .collect(),
+        )
+    }
+
+    pub fn from_message(message: &Message, base_asset_id: &AssetId) -> Self {
+        let retryable_flag_bytes = if message.is_retryable_message() {
+            RETRYABLE_BYTE
+        } else {
+            NON_RETRYABLE_BYTE
+        };
+        let address_bytes = message.recipient().as_ref();
+        let asset_id_bytes = base_asset_id.as_ref();
+        let amount_bytes = message.amount().to_be_bytes();
+        let nonce_bytes = message.nonce().as_slice();
+
+        Self(
+            retryable_flag_bytes
+                .iter()
+                .chain(address_bytes)
+                .chain(asset_id_bytes)
+                .chain(amount_bytes.iter())
+                .chain(nonce_bytes)
+                .copied()
+                .collect(),
+        )
+    }
+
+    fn from_slice(slice: &[u8]) -> Self {
+        Self(slice.into())
+    }
+
+    pub fn owner(&self) -> Address {
+        const ADDRESS_START: usize = RETRYABLE_FLAG_SIZE;
+        const ADDRESS_END: usize = ADDRESS_START + Address::LEN;
+        let address: [u8; Address::LEN] = self.0[ADDRESS_START..ADDRESS_END]
```

Comment:
And here

Reply:
See here: https://github.com/FuelLabs/fuel-core/pull/2463#discussion_r1889854801